cmd/src/batch_exec_json_lines_ui.go

```diff
@@ -0,0 +1,222 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/sourcegraph/src-cli/internal/batches"
+	"github.com/sourcegraph/src-cli/internal/batches/executor"
+	"github.com/sourcegraph/src-cli/internal/batches/graphql"
+	"github.com/sourcegraph/src-cli/internal/batches/workspace"
+)
+
+var _ batchExecUI = &batchExecJSONLinesUI{}
+
+type batchExecJSONLinesUI struct {
+}
+
+func (ui *batchExecJSONLinesUI) ParsingBatchSpec() {
+	logOperationStart("PARSING_BATCH_SPEC", "")
+}
+func (ui *batchExecJSONLinesUI) ParsingBatchSpecSuccess() {
+	logOperationSuccess("PARSING_BATCH_SPEC", "")
+}
+
+func (ui *batchExecJSONLinesUI) ParsingBatchSpecFailure(err error) {
+	logOperationFailure("PARSING_BATCH_SPEC", err.Error())
+}
+
+func (ui *batchExecJSONLinesUI) ResolvingNamespace() {
+	logOperationStart("RESOLVING_NAMESPACE", "")
+}
+func (ui *batchExecJSONLinesUI) ResolvingNamespaceSuccess(namespace string) {
+	logOperationSuccess("RESOLVING_NAMESPACE", fmt.Sprintf("Namespace: %s", namespace))
+}
+func (ui *batchExecJSONLinesUI) PreparingContainerImages() {
+	logOperationStart("PREPARING_DOCKER_IMAGES", "")
+}
+func (ui *batchExecJSONLinesUI) PreparingContainerImagesProgress(percent float64) {
+	logOperationProgress("PREPARING_DOCKER_IMAGES", fmt.Sprintf("%d%% done", int(percent*100)))
+}
+func (ui *batchExecJSONLinesUI) PreparingContainerImagesSuccess() {
+	logOperationSuccess("PREPARING_DOCKER_IMAGES", "")
+}
+func (ui *batchExecJSONLinesUI) DeterminingWorkspaceCreatorType() {
+	logOperationStart("DETERMINING_WORKSPACE_TYPE", "")
+}
+func (ui *batchExecJSONLinesUI) DeterminingWorkspaceCreatorTypeSuccess(wt workspace.CreatorType) {
+	switch wt {
+	case workspace.CreatorTypeVolume:
+		logOperationSuccess("DETERMINING_WORKSPACE_TYPE", "VOLUME")
+	case workspace.CreatorTypeBind:
+		logOperationSuccess("DETERMINING_WORKSPACE_TYPE", "BIND")
+	}
+}
+func (ui *batchExecJSONLinesUI) ResolvingRepositories() {
+	logOperationStart("RESOLVING_REPOSITORIES", "")
+}
+
+func (ui *batchExecJSONLinesUI) ResolvingRepositoriesDone(repos []*graphql.Repository, unsupported batches.UnsupportedRepoSet, ignored batches.IgnoredRepoSet) {
+	if unsupported != nil && len(unsupported) != 0 {
+		logOperationSuccess("RESOLVING_REPOSITORIES", fmt.Sprintf("%d unsupported repositories", len(unsupported)))
+	} else if ignored != nil && len(ignored) != 0 {
+		logOperationSuccess("RESOLVING_REPOSITORIES", fmt.Sprintf("%d ignored repositories", len(ignored)))
+	} else {
+		logOperationSuccess("RESOLVING_REPOSITORIES", fmt.Sprintf("Resolved %d repositories", len(repos)))
```

Comment:
Since this will presumably end up in the UI, these should all be pluralised. (Let's introduce or import a helper, since this comment is going to come up a bunch.)